README.md
# ⚓ ᏕiraDoc

[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![dependencies Status](https://status.david-dm.org/gh/SiraDoc/siradoc.svg)](https://david-dm.org/SiraDoc/siradoc)

Static documentation generator for scripting manual

## Demo

Looking for a demo? A simple demo is available at https://Sira.github.io/demo/!

<p align="center">
  <img src="./etc/demo.png"/>
</p>

## Installing

### NPM

To install SiraDoc using npm, execute the following command:

```
npm install -g @siradoc/siradoc
```

[![NPM](https://nodei.co/npm/@siradoc/siradoc.png)](https://www.npmjs.com/package/@siradoc/siradoc)

## TODO

- [ ] Add search capability
- [ ] Add test for CI

## License

This software uses other open-source components. For a full list, see the `LICENSE` file.

```
MIT License

Copyright (c) 2021 Jen-Chieh Shen

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.


## 0.3.0 (Unreleased)
> Released N/A

* N/A

## 0.2.0
> Released 29, 2021

* Update image `favicon.ico`.
* Fix CROS issue.

## 0.1.0
> Released Aug 29, 2021

* Initial release.


lib/cli.js
#!/usr/bin/env node
/**
 * $File: cli.js $
 * $Date: 2021-08-28 23:15:54 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

const fs = require('fs');
const sira = require('./sira');

const usage =
      "usage : sira I_FOLDER [output O_FOLDER] \n" +
      "\n" +
      "SiraDoc : Export static website for your Scripting Manual\n" +
      "\n" +
      "positional arguments:\n" +
      "  I_FOLDER          Directory with your makrdown files\n" +
      "\n" +
      "optional arguments:\n" +
      "  O_FOLDER          Output documentation in this directory\n";

/* CLI */
const cli_md = function (iPath, oPath) {
  iPath = iPath || "";
  oPath = oPath || "";

  // Check valid args.
  if (iPath == "")
    console.log(usage);
  // Check directory/file exists.
  else if (!fs.existsSync(iPath)) {
    console.log("Invalid build path for SiraDoc to build");
  }
  // Do build action.
  else {
    sira.build(iPath, oPath);
  }
};


if (require.main === module) {
  let args = process.argv;
  cli_md(args[2], args[3]);
}


lib/util.js
#!/usr/bin/env node
/**
 * $File: util.js $
 * $Date: 2021-08-29 13:32:24 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

/* Conversion Keywords */
const slashKey = "_sl_";
const spaceKey = "_sp_";  // Must be the same as server.

const obKey      = "_ob_";   // open bracket '('.
const cbKey      = "_cb_";   // close bracket ')'.
const ocbKey     = "_ocb_";  // open curly bracket '{'.
const ccbKey     = "_ccb_";  // close curly bracket '}'.
const osbKey     = "_osb_";  // open square bracket '['.
const csbKey     = "_csb_";  // close square bracket ']'.

const atKey      = "_at_";   // At key '@'.
const caretKey   = "_cr_";   // Caret key '^'.
const bqKey      = "_bq_";   // Back quote key '`'.
const tildeKey   = "_tl_";   // Tilde key '~'.
const hashKey    = "_hs_";   // Hash key '#'.
const dollarKey  = "_dl_";   // Dollar '$' key.
const percentKey = "_pc_";   // Percent '%' key.
const andKey     = "_and_";  // And '&' key.
const plusKey    = "_pl_";   // And '+' key.
const quoteKey   = "_qt_";   // Quote ' key.
const exclaimKey = "_ex_";   // Exclamation mark key '!'.

const periodKey    = "_pri_";  // Period '.' key.
const equalKey     = "_eq_";   // Equals ' =' key.
const commaKey     = "_cma_";  // Comma ',' key.
const semicolonKey = "_sc_";   // Semicolon ';' key.

/**
 * Apply conversion rules.
 * @param { string } rawStr : Unrefined string that have not apply
 * conversion rules.
 * @param { boolean } revert : Convert back.
 */
function applyConversionRule(rawStr, revert = false) {
  if (revert) {
    rawStr = rawStr.split(slashKey).join("/");
    rawStr = rawStr.split(spaceKey).join(" ");

    rawStr = rawStr.split(obKey).join("(");
    rawStr = rawStr.split(cbKey).join(")");
    rawStr = rawStr.split(ocbKey).join("{");
    rawStr = rawStr.split(ccbKey).join("}");
    rawStr = rawStr.split(osbKey).join("[");
    rawStr = rawStr.split(csbKey).join("]");

    rawStr = rawStr.split(atKey).join("@");
    rawStr = rawStr.split(bqKey).join("`");
    rawStr = rawStr.split(caretKey).join("^");
    rawStr = rawStr.split(tildeKey).join("~");
    rawStr = rawStr.split(hashKey).join("#");
    rawStr = rawStr.split(dollarKey).join("$");
    rawStr = rawStr.split(percentKey).join("%");
    rawStr = rawStr.split(andKey).join("&");
    rawStr = rawStr.split(plusKey).join("+");
    rawStr = rawStr.split(quoteKey).join("'");
    rawStr = rawStr.split(exclaimKey).join("!");

    rawStr = rawStr.split(periodKey).join(".");
    rawStr = rawStr.split(equalKey).join("=");
    rawStr = rawStr.split(commaKey).join(",");
    rawStr = rawStr.split(semicolonKey).join(";");
  } else {
    rawStr = rawStr.replace(/\//g, slashKey);
    rawStr = rawStr.replace(/ /g, spaceKey);

    rawStr = rawStr.replace(/\(/g, obKey);
    rawStr = rawStr.replace(/\)/g, cbKey);
    rawStr = rawStr.replace(/\{/g, ocbKey);
    rawStr = rawStr.replace(/\}/g, ccbKey);
    rawStr = rawStr.replace(/\[/g, osbKey);
    rawStr = rawStr.replace(/\]/g, csbKey);

    rawStr = rawStr.replace(/\@/g, atKey);
    rawStr = rawStr.replace(/\`/g, bqKey);
    rawStr = rawStr.replace(/\^/g, caretKey);
    rawStr = rawStr.replace(/\~/g, tildeKey);
    rawStr = rawStr.replace(/\#/g, hashKey);
    rawStr = rawStr.replace(/\$/g, dollarKey);
    rawStr = rawStr.replace(/\%/g, percentKey);
    rawStr = rawStr.replace(/\&/g, andKey);
    rawStr = rawStr.replace(/\+/g, plusKey);
    rawStr = rawStr.replace(/\'/g, quoteKey);
    rawStr = rawStr.replace(/\!/g, exclaimKey);

    rawStr = rawStr.replace(/\./g, periodKey);
    rawStr = rawStr.replace(/\=/g, equalKey);
    rawStr = rawStr.replace(/\,/g, commaKey);
    rawStr = rawStr.replace(/\;/g, semicolonKey);
  }
  return rawStr;
}

/*
 * Module Exports
 */
module.exports.slashKey = slashKey;
module.exports.applyConversionRule = applyConversionRule;


site/js/head.js
/**
 * $File: head.js $
 * $Date: 2018-09-29 01:44:08 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2018 by Shen, Jen-Chieh $
 */

"use strict";

/**
 * NOTE(jenchieh): The goal of this file is to minimize the HTML
 * code in all `index.html' files. So we can have consistence
 * style in each page.
 *
 * Mainly we do not have to change anything here, instead we will
 * often change the code in the `css-lib.html' file instead. :D
 */


/**
 * jQuery entry.
 */
(function ($) {
  let jsLib = $('#js-lib');
  let cssLib = $('#css-lib');

  /**
   * jQuery program entry.
   */
  function jQueryMain() {
    jsLib.load('../js-lib.html');
    cssLib.load('../css-lib.html');
  }
  jQueryMain();  // Execute the program entry.

}(this.jQuery));


site/js/main.js
/**
 * $File: main.js $
 * $Date: 2018-09-27 20:38:08 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2018 by Shen, Jen-Chieh $
 */

"use strict";

/**
 * jQuery entry.
 */
(function ($) {
  /* Collect all pages. */
  var manualPage = $('#manual-page');
  var scriptReferencePage = $('#script-reference-page');

  /* Collect Require Components. */
  var scrollBarTitle = $('#scroll-bar-title');

  /* Content */
  var content = $('#content');

  /* Buttons */

  /* Regular Decoration */
  var sbContainer = $('#scroll-bar-container');

  /* Search Input */
  var searchForm = $('#search-input-container form');
  var searchInputBorder = $('#search-input-border');
  var searchInputContainer = $('#search-input-container');
  var searchInput = $('#search-input-container input');
  var searchBtn = $('#search-btn');

  /* Search Result */
  var searchRes = null;

  //---------------------- Functions ---------------------------//

  /* Search Input */
  searchInputContainer.click(function () {
    searchInput.focus();
  });

  searchBtn.click(function () {
    doSearch();
  });


  // Prevent search default submit action.
  searchForm.submit(function(e) { e.preventDefault(); });

  searchInput.keypress(function (e) {
    if (e.which != 13)
      return;
    doSearch();
  });

  /**
   * Do the search.
   */
  function doSearch() {
    let searchKeyword = searchInput.val().trim();

    // If search for nothing just return it.
    if (searchKeyword == "")
      return;

    /* IMPORTANT: Apply conversion rule. */
    // This rule must match the server side.
    searchKeyword = searchKeyword.replace(/ /g, spaceKey);

    // Load to search page.
    addParamToURL('search', searchKeyword, true);
  }

  function showChildren(obj) { obj.children().show(); }
  function hideChildren(obj) { obj.children().hide(); }

  /* Open the SB directory and turn the arrow text off. */
  function openSBDir(items, arrow) {
    showChildren(items);
    arrow.text("-");
  }

  /* Close the SB directory and turn the arrow text on. */
  function closeSBDir(items, arrow) {
    hideChildren(items);
    arrow.text("+");
  }

  /* Toggle the SB directory and turn the arrow text on/off. */
  function toggleSBDir(items, arrow) {
    if (arrow.text() == "+")
      openSBDir(items, arrow);
    else
      closeSBDir(items, arrow);
  }


  /* Register button event */
  function addSBDirButtonEvent() {
    let sbDir = $('.sb-dir');
    let arrows = $('.arrow');
    let arrowsText = $('.arrow + span');

    arrows.click(function (e) {
      // Stop overlaping `li' tag's click event trigger.
      e.stopPropagation();

      let items = $(this).siblings('ul');

      toggleSBDir(items, $(this));
    });

    arrowsText.click(function (e) {
      // Stop overlaping `li' tag's click event trigger.
      e.stopPropagation();

      let items = $(this).siblings('ul');
      let arrow = $(this).siblings('.arrow');

      toggleSBDir(items, arrow);
    });

    let currentContentPage = getUrlParameter('page');

    let pathDir = [];

    if (currentContentPage !== null)
      pathDir = currentContentPage.split(slashKey);

    let dirLayer = 0;
    let currentPathDir = pathDir[dirLayer];

    // Initialize by URL param.
    sbDir.each(function () {
      let items = $(this).find('ul');
      let arrow = $(this).find('.arrow');

      let path = $(this).attr('id');

      if (currentPathDir == path) {
        openSBDir(items, arrow);

        // Add up directory layer.
        ++dirLayer;

        // Setup the next directory tree.
        currentPathDir += slashKey + pathDir[dirLayer];
      } else {
        // Close the directory as default.
        closeSBDir(items, arrow);
      }
    });
  }

  function addSBFileButtonEvent() {
    let sbFile = $('.sb-file');

    sbFile.click(function (e) {
      // Stop overlaping `div' tag's click event trigger.
      e.stopPropagation();

      let contentPage = $(this).attr('id');

      addParamToURL("page", contentPage, true);
    });

    let currentContentPage = getUrlParameter('page');

    let selectedFilename = [];

    if (currentContentPage != null) {
      selectedFilename = currentContentPage.split(slashKey);
      selectedFilename = selectedFilename[selectedFilename.length - 1];
    }

    // Check if current file selected. Highlight it!
    sbFile.each(function () {
      let filePath = $(this).attr('id');

      let filename = filePath.split(slashKey);
      filename = filename[filename.length - 1];

      // Found the selected file?
      if (selectedFilename == filename) {
        $(this).addClass('sb-file-selected');

        // Scroll to that file selected.
        sbContainer.animate({
          /*
           * NOTE: 100 is the height from the header, and plus 10 just
           * to offset a bit more to look better.
           */
          scrollTop: $(this).offset().top - 110
        }, 500);
      }
    });
  }

  /**
   * Add a parameter to current URL.
   *
   * @param { string } paramName : parameter name.
   * @param { string } paramValue : parameter value.
   * @param { boolean } clean : Clean param?
   */
  function addParamToURL(paramName, paramValue, clean) {
    let url = document.location.href;

    // Remove all parameters?
    if (clean)
      url = url.split('?')[0];

    if (url.indexOf('?') != -1) {
      url += "&";
    } else {
      url += "?";
    }

    url += paramName + "=" + paramValue;

    // Set URL and reload the page.
    document.location = url;
  }

  function applyConfig() {
    /* Reload possible changing variables. */
    let manualName = $('.manual-name');
    let copyright = $('.copyright');
    let homepageLink = $('.homepage-link');

    let versionTitle01 = $('.version-title-01');
    let versionTitle02 = $('.version-title-02');
    let versionTitle03 = $('.version-title-03');
    let versionTitle04 = $('.version-title-04');
    let versionTitle05 = $('.version-title-05');
    let versionTitle06 = $('.version-title-06');
    let versionTitle07 = $('.version-title-07');

    let versionNum01 = $('.version-num-01');
    let versionNum02 = $('.version-num-02');
    let versionNum03 = $('.version-num-03');
    let versionNum04 = $('.version-num-04');
    let versionNum05 = $('.version-num-05');
    let versionNum06 = $('.version-num-06');
    let versionNum07 = $('.version-num-07');

    manualName.text(manual_name);
    copyright.text(copyright_text);
    homepageLink.text(homepage_text);
    homepageLink.attr('href', homepage_url);

    versionTitle01.text(version_title_01);
    versionTitle02.text(version_title_02);
    versionTitle03.text(version_title_03);
    versionTitle04.text(version_title_04);
    versionTitle05.text(version_title_05);
    versionTitle06.text(version_title_06);
    versionTitle07.text(version_title_07);

    versionNum01.text(version_num_01);
    versionNum02.text(version_num_02);
    versionNum03.text(version_num_03);
    versionNum04.text(version_num_04);
    versionNum05.text(version_num_05);
    versionNum06.text(version_num_06);
    versionNum07.text(version_num_07);
  }
  applyConfig();

  /**
   * Apply customizable color.
   */
  function applyTheme() {
    let header = $('header');
    header.css('background-color', header_color);

    let arrows = $('.arrow');
    arrows.css('background-color', arrow_color);

    let th = $('th');
    th.css('background-color', th_color);

    if (th_show == false) {
      let thead = $('thead');
      thead.css('display', 'none');
    }
  }
  applyTheme();

  function loadContent() {
    let url = window.location.href;
    let type = basePath(window.location.pathname);
    let base = '..' + type;
    let docBase = '';

    if (type === '/Manual') {
      document.title = manual_name + ' - Scripting Manual';
      scrollBarTitle.text(manual_name + " Manul");
      searchInput.attr('placeholder', si_manual_placeholder);
      docBase = '/doc';
    } else {
      document.title = manual_name + ' - Scripting API';
      scrollBarTitle.text("Scripting API");
      searchInput.attr('placeholder', si_api_placeholder);
      docBase = '/api';
    }

    let page = getUrlParameter('page');
    if (page !== null)
      page = applyConversionRule(page, true);
    else {
      docBase = "";
      page = "intro";
    }

    let docPath = base + docBase + '/' + page + '.html';

    $.get(docPath, function(result) {
      content.html(result);

      addSBDirButtonEvent();
      addSBFileButtonEvent();
    });
  }
  loadContent();

}(this.jQuery));


/**
 * Get URL parameter.
 *
 * SOURCE: https://stackoverflow.com/questions/19491336/get-url-parameter-jquery-or-how-to-get-query-string-values-in-js
 * @param { string } paramName : name of the parameter.
 */
function getUrlParameter(paramName) {
  let sPageURL = decodeURIComponent(window.location.search.substring(1));
  let sURLVariables = sPageURL.split('&');
  for (let index = 0; index < sURLVariables.length; ++index) {
    let sParameterName = sURLVariables[index].split('=');
    if (sParameterName[0] === paramName) {
      return sParameterName[1] === undefined ? true : sParameterName[1];
    }
  }
  return null;
}

/**
 * Clean all URL parameters.
 */
function cleanParamFromURL() {
  let url = document.location.href;

  let splitUrl = url.split('?');
  url = splitUrl[0];

  // Make sure there are not param after url.
  if (splitUrl.length == 2)
    document.location = url;
}

/** Return base of the url/path. */
function basePath(path) { return path.replace(/\/[^\/]+\/?$/, ''); }


site/js/config.js
/**
 * $File: config.js $
 * $Date: 2018-09-28 22:33:54 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2018 by Shen, Jen-Chieh $
 */

"use strict";

/* Manual */
const manual_name = "[Manual Name]";

/* Theme */
const header_color = '#222C37';
const arrow_color  = '#19E3B1';
const th_show      = true;       // table header
const th_color     = '#A2F2DE';  // table header color

/* Version */
const version_title_01 = "[Manual] Version: ";
const version_title_02 = "";
const version_title_03 = "";
const version_title_04 = "";
const version_title_05 = "";
const version_title_06 = "";
const version_title_07 = "";

const version_num_01 = "0.0.1";
const version_num_02 = "";
const version_num_03 = "";
const version_num_04 = "";
const version_num_05 = "";
const version_num_06 = "";
const version_num_07 = "";

/* Copyright */
const copyright_text = "Copyright © [yyyy] [Group Name]. Built: [yyyy-mm-dd].";

/* Others */
const intro_content  = 'intro';   // Default content page.
const search_content = 'search';  // Search page.

/* Homepage Link */
const homepage_url  = "https://github.com/jcs090218/SiraDoc";  // Link to your homepage.
const homepage_text = "SiraDoc.github";                                   // Text represent your homepage.

/* Search Input Text */
const si_manual_placeholder = "Search manual...";
const si_api_placeholder    = "Search scripting...";

const si_input_size = 17;  // character unit.


site/js/showdown-highlight.js
/**
 * $File: showdown-highlight.js $
 * $Date: 2020-01-13 17:30:44 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2020 by Shen, Jen-Chieh $
 */

"use strict";

// SEE: https://github.com/showdownjs/showdown/issues/215

showdown.extension('codehighlight', function() {
  function htmlunencode(text) {
    return (
      text
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
    );
  }
  return [
    {
      type: 'output',
      filter: function (text, converter, options) {
        // use new shodown's regexp engine to conditionally parse codeblocks
        var left  = '<pre><code\\b[^>]*>',
            right = '</code></pre>',
            flags = 'g',
            replacement = function (wholeMatch, match, left, right) {
              // unescape match to prevent double escaping
              match = htmlunencode(match);
              return left + hljs.highlightAuto(match).value + right;
            };
        return showdown.helper.replaceRecursiveRegExp(text, replacement, left, right, flags);
      }
    }
  ];
});


lib/sira.js
#!/usr/bin/env node
/**
 * $File: sira.js $
 * $Date: 2021-08-28 23:16:25 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2021 by Shen, Jen-Chieh $
 */

"use strict";

/*
 * Dependencies
 */
const fs = require("fs");
const path = require('path');

const fse = require('fs-extra');
const dirTree = require('directory-tree');
const replaceExt = require('replace-ext');

const HTMLParser = require('node-html-parser');
const showdown = require('showdown');
const showdownHighlight = require("showdown-highlight");

const util = require('./util');

/* Global Variables */
var layerNum = -1;

/*
 * Util
 */
function info(msg) { console.log("[INFO] " + msg); }
function error(msg) { console.log("[ERROR] " + msg); }

function mkdir(path) { fs.mkdirSync(path, { recursive: true }); }
function readFile(path) { return fs.readFileSync(path, 'utf8'); }
function writeFile(path, content) { fs.writeFileSync(path, content, { encoding: 'utf8' }); }

/*
 * Core
 */

/* Create index with directory. */
function createIndexWithDir(dir, inParent) {
  let ul = HTMLParser.parse('<ul></ul>');
  inParent.appendChild(ul);

  let parent = inParent.querySelector('ul');

  ++layerNum;

  let currentDir = "";

  for (let index = 0; index < dir.length; ++index) {
    let pathObj = dir[index];

    if (pathObj.path.charAt(0) != "/")
      currentDir = pathObj.path;

    let sbType = "sb-dir";
    if (pathObj.type == "file") {
      sbType = "sb-file";
    }

    let isDir = (pathObj.type != "file");

    let newPath = pathObj.path;
    newPath = newPath.substring(1);  // remove first slash in path
    if (!isDir) newPath = newPath.replace(/\.[^/.]+$/, "");  // Remove extension if file.

    /* IMPORTANT: Apply conversion rule. */
    newPath = util.applyConversionRule(newPath);

    let dirOrFileName = pathObj.name;
    dirOrFileName = dirOrFileName.replace(/\.[^/.]+$/, "");  // Remove extension if there is.

    let li = HTMLParser.parse("<li class=" + sbType +" id=" + newPath + "></li>");
    parent.appendChild(li);

    let newPathNode = parent.querySelector('#' + newPath);

    let htmlDirOrFileName = "<span>" + dirOrFileName + "</span>";
    if (isDir) {
      htmlDirOrFileName = "<div class='arrow'>+</div>" + htmlDirOrFileName;
    }

    let temp = HTMLParser.parse(htmlDirOrFileName);
    newPathNode.appendChild(temp);

    newPathNode.classList.add("sb-layer-" + layerNum);

    if (pathObj.children != null && pathObj.children.length != 0) {
      createIndexWithDir(pathObj.children, newPathNode);
    }
  }

  --layerNum;

  return inParent;
}

function prepareSite(src, dest) {
  info('Copying template to current directory...');
  let sitePath = path.join(__dirname, '..', 'site');
  fse.copySync(sitePath, dest);
}

function getDocTree(path, ext) {
  const tree = dirTree(path, { extensions: ext, normalizePath: true });
  sortTreeByType(tree.children, 'directory');
  var removePath = path;
  removePath = removePath.replace("./", "");
  removePath = removePath.replace(/\\/g, '/');
  removeLeadPath(tree.children, removePath);
  return tree;
}

function md2html(markdown) {
  showdown.setFlavor('github');
  let converter = new showdown.Converter({ extensions: [showdownHighlight({ pre: true })]});
  return converter.makeHtml(markdown);
}

function genHtmlFromMd(src, dest) {
  let markdown = readFile(src);
  let html = md2html(markdown);
  writeFile(dest, html);
}

function buildHtmlFiles(tree, src, dest) {
  for (let index = 0; index < tree.length; ++index) {
    let pathObj = tree[index];
    if (pathObj.children != null && pathObj.children.length != 0) {
      buildHtmlFiles(pathObj.children, src, dest);
    }

    if (pathObj.type === 'file') {
      let destFile = path.join(dest, replaceExt(pathObj.path, '.html'));
      let destDir = path.dirname(destFile);
      mkdir(destDir);

      let srcFile = path.join(src, pathObj.path);
      genHtmlFromMd(srcFile, destFile);
    }
  }
}

function buildIntro(src, dest) {
  info('Building intro pages...');

  let srcIntroManual = path.join(src, 'Manual.md');
  let srcIntroScriptRef = path.join(src, 'ScriptReference.md');

  let destIntroManual = path.join(dest, 'Manual', 'intro.html');
  let destIntroScriptRef = path.join(dest, 'ScriptReference', 'intro.html');

  genHtmlFromMd(srcIntroManual, destIntroManual);
  genHtmlFromMd(srcIntroScriptRef, destIntroScriptRef);
}

function buildDocs(src, dest) {
  info('Building documents directory...');

  let pathManual = path.join(src, 'Manual');
  let pathScriptRef = path.join(src, 'ScriptReference');
  let treeManual = getDocTree(pathManual, /\.md/);
  let treeScriptRef = getDocTree(pathScriptRef, /\.md/);

  let oPathManual = path.join(dest, 'Manual', 'doc');
  let oPathScriptRef = path.join(dest, 'ScriptReference', 'api');

  buildHtmlFiles(treeManual.children, pathManual, oPathManual);
  buildHtmlFiles(treeScriptRef.children, pathScriptRef, oPathScriptRef);
}

function buildIndex(indexPath, docPath) {
  let html = readFile(indexPath);
  let tree = getDocTree(docPath, /\.html/);

  let root = HTMLParser.parse(html);
  let indexPos = root.querySelector('#index-pos');

  let newPage = createIndexWithDir(tree.children, indexPos);
  indexPos.replaceWith(newPage);
  writeFile(indexPath, root.toString());
}

function buildSidebar(src, dest) {
  let indexManual = path.join(dest, 'Manual', 'index.html');
  let indexScriptRef = path.join(dest, 'ScriptReference', 'index.html');

  let oPathManual = path.join(dest, 'Manual', 'doc');
  let oPathScriptRef = path.join(dest, 'ScriptReference', 'api');

  buildIndex(indexManual, oPathManual);
  buildIndex(indexScriptRef, oPathScriptRef);
}

function buildSite(src, dest) {
  info('Generating static site...');

  buildSidebar(src, dest);  // build index links on the side
}

function build(iPath, oPath) {
  let src = path.join(process.cwd(), iPath);
  let dest = path.join(process.cwd(), oPath);

  prepareSite(src, dest);  // copy template over
  buildIntro(src, dest);
  buildDocs(src, dest);
  buildSite(src, dest);
}

/**
 * Sort the tree result by directory/file type.
 * @param { tree.children } tree : Tree children.
 * @param { string } type : Sort type, enter 'directory' or 'file'.
 */
function sortTreeByType(tree, type = 'directory') {
  if (type != 'directory' && type != 'file')
    return;

  let tarList = [];  // target list.
  let anoList = [];  // another list.

  /* Split path object into two arrays by type. */
  for (let index = 0; index < tree.length; ++index) {
    let pathObj = tree[index];
    if (pathObj.children != null && pathObj.children.length != 0) {
      sortTreeByType(pathObj.children, type);
    }

    // Add path object by type.
    if (pathObj.type == type)
      tarList.push(pathObj);
    else
      anoList.push(pathObj);
  }

  /* Copy array over. */
  let resultTree = tarList.concat(anoList);

  for (let index = 0; index < tree.length; ++index) {
    tree[index] = resultTree[index];
  }
}

/**
 * Remove the `MANUAL_DIR_PATH' or `API_DIR_PATH', so when the client
 * receive the data would not need to care where is the api directory located.
 * @param { JSON } dir : directory JSON object.
 * @param { typename } rmPath : Param desc here..
 */
function removeLeadPath(dir, rmPath) {
  for (let index = 0; index < dir.length; ++index) {
    let pathObj = dir[index];

    if (pathObj.children != null && pathObj.children.length != 0) {
      removeLeadPath(pathObj.children, rmPath);
    }

    // Remove the `MANUAL_DIR_PATH' or `API_DIR_PATH' path.
    pathObj.path = pathObj.path.replace(rmPath, "");
  }
}

/*
 * Module Exports
 */
module.exports.build = build;
